---
title: "CD_EGC analysis final"
author: "Ryan Brown"
date: "2024-03-05"
output: html_document
---

## Load Packages
```{r Load packages}
library(dplyr)
library(Seurat)
library(cowplot)
library(patchwork)
library(ggplot2)
library(clusterProfiler)
library(enrichR)
library(enrichplot)
library(grid)
library(rstatix)
library(readr)
library(tidyr)
library(topGO)
library(org.Hs.eg.db)
library(EnhancedVolcano)
```

# Citation for dataset:
# Kong, L., Pokatayev, V., Lefkovith, A., Carter, G. T., Creasey, E. A., Krishna, C., Subramanian, S., Kochar, B., Ashenberg, O., Lau, H., Ananthakrishnan, A. N., Graham, D. B., Deguine, J., &amp; Xavier, R. J. (2023). The landscape of immune dysregulation in crohnâ€™s disease revealed through single-cell transcriptomic profiling in the ileum and colon. Immunity, 56(2). https://doi.org/10.1016/j.immuni.2023.01.002 

```{r Load data and create seurat object}
# Load data
CD_Colon_stroma_10X <- Read10X("CD_Colon_10X_data/")

# Create Seurat object using GEX matrices
CD_Colon_stroma.Seurat <- CreateSeuratObject(counts = CD_Colon_stroma_10X, project = 'CD_Colon_stroma')
# Explore seurat object metadata
CD_Colon_stroma.Seurat
head(colnames(CD_Colon_stroma.Seurat))
table(CD_Colon_stroma.Seurat$orig.ident)
head(CD_Colon_stroma.Seurat@meta.data)

# Add Condition column w/ disease state - Healthy vs. Crohn's
CD_Colon_stroma.Seurat[["Condition"]] <- case_when(
  startsWith(as.character(CD_Colon_stroma.Seurat$orig.ident), "H") ~ "Healthy",
  startsWith(as.character(CD_Colon_stroma.Seurat$orig.ident), "N") ~ "Crohn's",
  startsWith(as.character(CD_Colon_stroma.Seurat$orig.ident), "I") ~ "Crohn's")
# Check
CD_Colon_stroma.Seurat[["Condition"]] %>% 
  slice_sample(n=10)

# Add Subcondition column w/ disease state - Inflamed vs. Noninflamed
CD_Colon_stroma.Seurat[["Subcondition"]] <- case_when(
  startsWith(as.character(CD_Colon_stroma.Seurat$orig.ident), "H") ~ "Healthy",
  startsWith(as.character(CD_Colon_stroma.Seurat$orig.ident), "N") ~ "CD_Noninflamed",
  startsWith(as.character(CD_Colon_stroma.Seurat$orig.ident), "I") ~ "CD_Inflamed")
# Check
CD_Colon_stroma.Seurat[["Subcondition"]] %>% 
  slice_sample(n=10)

# Change idents to condition
Idents(CD_Colon_stroma.Seurat) <- "Condition"
# Check
head(Idents(CD_Colon_stroma.Seurat))
# Change order
CD_Colon_stroma.Seurat$Condition <- factor(CD_Colon_stroma.Seurat$Condition, levels = c("Healthy", "Crohn's"))
```

```{r PCA}
# Normalize data
CD_Colon_stroma.Seurat <- NormalizeData(CD_Colon_stroma.Seurat)

# Find variable features
CD_Colon_stroma.Seurat <- FindVariableFeatures(CD_Colon_stroma.Seurat, selection.method = "vst", nfeatures = 2000)

# Scale the data
all.genes <- rownames(CD_Colon_stroma.Seurat)
CD_Colon_stroma.Seurat <- ScaleData(CD_Colon_stroma.Seurat, features = all.genes)

# Run PCA
CD_Colon_stroma.Seurat <- RunPCA(CD_Colon_stroma.Seurat) 
# Print top 5 PCs
print(CD_Colon_stroma.Seurat[["pca"]], dims = 1:5, nfeatures = 5)
# Visualize with DimHeatmap
DimHeatmap(CD_Colon_stroma.Seurat, dims = 1:5, cells = 500, balanced = TRUE)

# Save because this step takes a long time
#saveRDS(CD_Colon_stroma.Seurat, file = "CD_Colon_stroma.Seurat.RDS")
```

```{r UMAP}
# Read RDS file if starting from here
#CD_Colon_stroma.Seurat <- readRDS("CD_Colon_stroma.Seurat.RDS")

# ElbowPlot to determine the dimensionality of the dataset
ElbowPlot(CD_Colon_stroma.Seurat)

# Cluster the cells
# K-Nearest Neighbor graph
CD_Colon_stroma.Seurat <- FindNeighbors(CD_Colon_stroma.Seurat, dims = 1:8)
# Find Clusters with 0.5 resolution for first run through
CD_Colon_stroma.Seurat <- FindClusters(CD_Colon_stroma.Seurat, resolution = 0.46)
# Look at cluster IDs of the first 5 cells
head(Idents(CD_Colon_stroma.Seurat), 5)

# UMAP
CD_Colon_stroma.Seurat <- RunUMAP(CD_Colon_stroma.Seurat, dims = 1:8)

CD_Colon_stroma_UMAP <- DimPlot(CD_Colon_stroma.Seurat, reduction = "umap", group.by = "Condition")
CD_Colon_stroma_UMAP_labeled <- DimPlot(CD_Colon_stroma.Seurat, reduction = "umap", label = T, label.box = T, label.size = 8, repel = T) + theme(axis.line = element_line(size = 1), axis.text = element_text(size = 12), axis.title = element_text(size = 20)) + xlab("UMAP 1") + ylab("UMAP 2") + NoLegend()
CD_Colon_stroma_UMAP_labeled


plot_grid(CD_Colon_stroma_UMAP  + labs(title = "CD_Colon"), CD_Colon_stroma_UMAP_labeled)

# View conditions side by side
DimPlot(CD_Colon_stroma.Seurat, reduction = "umap", split.by = "Condition", label = TRUE) 

```

```{r Cluster ID}
## check for markers
# Find markers for every cluster compared to all remaining cells, report only the positive ones
CD_Colon_stroma.markers <- FindAllMarkers(CD_Colon_stroma.Seurat, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
# Print top two markers for each cluster
CD_Colon_stroma.markers %>% 
  group_by(cluster) %>% 
  top_n(n = 2, wt = avg_log2FC) %>% 
  print(n =36)

# Activated fibroblasts
VlnPlot(CD_Colon_stroma.Seurat, features = c("ADAMDEC1", "CCL19"))
  # Cluster 0
# Inflammatory fibroblasts
VlnPlot(CD_Colon_stroma.Seurat, features = c("IL11", "CHI3L1"))
  # Cluster 14
#  Fibroblasts 
VlnPlot(CD_Colon_stroma.Seurat, features = c("PDPN", "PDGFRA"))
  # Cluster 3,4,5,10,11


# Endothelial cells
VlnPlot(CD_Colon_stroma.Seurat, features = c("CD34"))
  # Clusters 6,9,16
# Lymphatic endothelial cells
VlnPlot(CD_Colon_stroma.Seurat, features = c("LYVE1", "CCL21"))
  # Cluster 1

# EGCs
VlnPlot(CD_Colon_stroma.Seurat, features = c("PLP1", "S100B"))
  # Cluster 8

# GREM1+ Myofibroblasts
VlnPlot(CD_Colon_stroma.Seurat, features = c("GREM1", "GREM2"))
  # Cluster 12
# HHIP+ Myofibroblasts
VlnPlot(CD_Colon_stroma.Seurat, features = c("HHIP", "NPNT"))
  # Cluster 7

# HIGD1B+ Pericytes
VlnPlot(CD_Colon_stroma.Seurat, features = c("CSPG4", "PDGFRB"))
  # Cluster 13, 15


# Stromal Cycling Cells
VlnPlot(CD_Colon_stroma.Seurat, features = c("HMGB2", "CDC20"))
  # Cluster 


CD_Seurat_cluster.ids <- c("Activated Fibroblasts", "Lymphatic Endothelial Cells", "Stromal Cycling Cells", "Fibroblasts 1", "Fibroblasts 2", "Fibroblasts 3", "Endothelial Cells 1", "Enteric Glial Cells", "Myofibroblasts 1", "Endothelial Cells 2", "Fibroblasts 4", "Fibroblasts 5", "Myofibroblasts 2", "Pericytes 1", "Inflammatory Fibroblasts", "Pericytes 2", "Endothelial Cells 3", "Endothelial Cells 4")
names(CD_Seurat_cluster.ids) <- levels(CD_Colon_stroma.Seurat)
CD_Colon_stroma.Seurat <- RenameIdents(CD_Colon_stroma.Seurat, CD_Seurat_cluster.ids)

CD_UMAP <- DimPlot(CD_Colon_stroma.Seurat, reduction = "umap", label = TRUE, label.box = T, label.size = 8, repel = T, cols = c('Activated Fibroblasts' = 'chocolate', 'Lymphatic Endothelial Cells' = 'khaki3', 'Stromal Cycling Cells' = 'slateblue1', 'Fibroblasts 1' = 'orange', 'Fibroblasts 2' = 'red', 'Fibroblasts 3' = 'orchid3', 'Endothelial Cells 1' = 'darkgoldenrod1', 'Myofibroblasts 1' = 'brown2', 'Enteric Glial Cells' = 'lawngreen', 'Endothelial Cells 2' = 'olivedrab', 'Fibroblasts 4' = 'lightseagreen', 'Fibroblasts 5' = 'yellow2', 'Myofibroblasts 2' = 'deepskyblue2', 'Pericytes 1' = 'coral', 'Inflammatory Fibroblasts' = 'firebrick2', 'Pericytes 2' = 'coral3', 'Endothelial Cells 2' = 'darkolivegreen3', 'Endothelial Cells 3' = 'skyblue2', 'Endothelial Cells 4' = 'deepskyblue3')) + theme(axis.line = element_line(size = 1), axis.text = element_text(size = 20), axis.title = element_text(size = 20), axis.ticks = element_line(size = 1), axis.ticks.length = unit(.25, "cm")) + xlab("UMAP 1") + ylab("UMAP 2") + NoLegend()
CD_UMAP
```

```{r EGC subcluster}
# EGC markers
VlnPlot(CD_Colon_stroma.Seurat, features = c("S100B", "PLP1"))
  # Cluster 8 are EGCs as they coexpress S100B and PLP1

# Subset cluster 8
CD_Colon_stroma_EGCs = subset(CD_Colon_stroma.Seurat, idents = "Enteric Glial Cells")
CD_Colon_stroma_EGCs = subset(CD_Colon_stroma_EGCs, PLP1 > 1 & S100B > 1)
  # Check cell # /condition
  table(CD_Colon_stroma_EGCs$Condition)
# Vlnplot
Idents(CD_Colon_stroma_EGCs) <- "Condition"

EGC_CD_Vln <- VlnPlot(CD_Colon_stroma_EGCs, features = c("S100B", "PLP1"), split.by = "Condition", cols = c("#232D4B", "#E57200")) & theme(axis.line = element_line(size = 1), axis.text = element_text(size = 20), title = element_text(size = 20), axis.title.y = element_text(size = 20), axis.title.x = element_blank(), axis.text.x = element_blank(), axis.ticks.y = element_line(size = 1), axis.ticks.length.y = unit(.25, "cm"), axis.ticks.x = element_blank(), plot.title = element_text(face = 'bold.italic'))
EGC_CD_Vln

## Subclustering 
# Reintegrate
# Make sure default assay is RNA
DefaultAssay(CD_Colon_stroma_EGCs) <- "RNA"
# Split and integrate
CD_EGCs.list = SplitObject(CD_Colon_stroma_EGCs, split.by = "Condition")
# Normalize and find variable features on list object
CD_EGCs.list <- lapply(X = CD_EGCs.list, FUN =  function(x) {
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})
# Select features that are repeatedly variable across datasets for integration
CD_EGCs_integration_features <- SelectIntegrationFeatures(object.list = CD_EGCs.list)
# Find integration anchors
CD_EGCs_integration_anchors <- FindIntegrationAnchors(object.list = CD_EGCs.list, anchor.features = CD_EGCs_integration_features)
# Integrate
CD_EGCs.integrated <- IntegrateData(anchorset = CD_EGCs_integration_anchors)
# Check default assay
DefaultAssay(CD_EGCs.integrated)

# Scale Data
CD_EGCs.integrated <- ScaleData(CD_EGCs.integrated)
# Run PCA
CD_EGCs.integrated <- RunPCA(CD_EGCs.integrated)
# ElbowPlot to determine the dimensionality of the dataset
ElbowPlot(CD_EGCs.integrated)

# Cluster the cells
# K-Nearest Neighbor graph
CD_EGCs.integrated <- FindNeighbors(CD_EGCs.integrated, dims = 1:5)
# Find Clusters with 0.5 resolution for first run through
CD_EGCs.integrated <- FindClusters(CD_EGCs.integrated, resolution = 0.5)

# UMAP
CD_EGCs.integrated <- RunUMAP(CD_EGCs.integrated, dims = 1:5)
CD_EGCs.integrated_UMAP_grouped <- DimPlot(CD_EGCs.integrated, reduction = "umap", group.by = "Condition")
CD_EGCs.integrated_UMAP_labeled <- DimPlot(CD_EGCs.integrated, reduction = "umap", label = TRUE)
plot_grid(CD_EGCs.integrated_UMAP_grouped  + labs(title = "IBD_10x_EGCs"), CD_EGCs.integrated_UMAP_labeled)
# View conditions side by side
DimPlot(CD_EGCs.integrated, reduction = "umap", split.by = "Condition", label = TRUE) 

```


```{r}
# Set RNA default assay, idents to condition
DefaultAssay(CD_EGCs.integrated) <- "RNA"
Idents(CD_EGCs.integrated) <- "Condition"
  # Check
  Idents(CD_EGCs.integrated)
  
  # Join layers
CD_EGCs.integrated <-JoinLayers(CD_EGCs.integrated)
# DE testing using findmarkers
CDvsHC_DE <- FindMarkers(CD_EGCs.integrated, ident.1 = "Crohn's", ident.2 = "Healthy")
CDvsHC_DE

# Add column for gene symbol
CDvsHC_DE$genesymbol <- rownames(CDvsHC_DE)
# Add column for entrez ID
CDvsHC_DE$entrezID <- mapIds(org.Hs.eg.db, keys = CDvsHC_DE$genes, column = "ENTREZID", keytype = "SYMBOL")
# Drop rows that don't have mapped entrez gene id (mostly mitochondrial_)
CDvsHC_DE <- CDvsHC_DE %>% drop_na()

# Volcano plot
Healthy_Crohns_Volcano <- EnhancedVolcano(CDvsHC_DE, lab = rownames(CDvsHC_DE), x = 'avg_log2FC', y = 'p_val_adj', FCcutoff = 1, labFace = "bold.italic", labSize = 5, legendLabSize = 20, axisLabSize = 20, raster = T, drawConnectors = T, title = NULL, subtitle = NULL, caption = NULL, gridlines.major = F, gridlines.minor = F) & theme(axis.line = element_line(size = 1), axis.text.x = element_text(size = 20), axis.text.y = element_text(size = 20), axis.ticks = element_line(size = 1), axis.ticks.length = unit(.25, "cm"), legend.position = 'top')
Healthy_Crohns_Volcano
```


````{r}
## Subset positively DE genes by avg_log2_FC >1 & p_val_adj <.05 and remove NAs
CD_upregulated <- filter(CDvsHC_DE, avg_log2FC > 1 & p_val_adj < 0.05)
# Check
CD_upregulated

# enrichGO
CD_upregulated_enrichGO <- enrichGO(gene = CD_upregulated$entrezID,
                            OrgDb = org.Hs.eg.db,
                            ont = "BP",
                            pAdjustMethod = "BH",
                            pvalueCutoff = 0.01,
                            qvalueCutoff = 0.05,
                            readable = T)
CD_upregulated_enrichGO
head(CD_upregulated_enrichGO)

# Filter to GO level 4
CD_upregulated_enrichGO <- gofilter(CD_upregulated_enrichGO, level = 4)
# Simplify repetitice GO terms
CD_upregulated_enrichGO <- simplify(CD_upregulated_enrichGO, cutoff =0.7, by ="p.adjust", select_fun=min)
# Bar plot
barplot(CD_upregulated_enrichGO, showCategory = 10)

mutate(CD_upregulated_enrichGO, qscore = -log(p.adjust, base = 10)) %>% 
  barplot(x="qscore", showCategory = 5) 


# Dot plot
CD_upregulated_enrichGO_dotplot <- dotplot(CD_upregulated_enrichGO, showCategory = 5)
CD_upregulated_enrichGO_dotplot <- CD_upregulated_enrichGO_dotplot & theme(rect = element_rect(linewidth = 1, color = 'black'), axis.line = element_line(size = 1), axis.text.y = element_text(size = 14), title = element_text(size = 20), axis.title.y = element_text(size = 20), axis.title.x = element_text(size = 20), axis.text.x = element_text(size = 20), axis.ticks.y = element_line(size = 1), axis.ticks.length.y = unit(.25, "cm"), legend.text = element_text(size = 14)) 
CD_upregulated_enrichGO_dotplot
```

```{r}
# AGPP DE
AGPP_KEGG_genes <- read.csv("AGPP_genes.csv")
AGPP_KEGG_vector <- AGPP_KEGG_genes$Gene
Idents(CD_EGCs.integrated) <- "Condition"
CD_EGCs_AGPP_KEGG <- subset(CD_EGCs.integrated, features = AGPP_KEGG_vector)
CD_EGCs_AGPP_KEGG <- ScaleData(CD_EGCs_AGPP_KEGG)

DefaultAssay(CD_EGCs_AGPP_KEGG) <- "RNA"

AGPP_KEGG_Markers <- FindMarkers(CD_EGCs_AGPP_KEGG, ident.1 = "Healthy", ident.2 = "Crohn's")
AGPP_KEGG_Markers
AGPP_KEGG_Markers <- subset(AGPP_KEGG_Markers, AGPP_KEGG_Markers$p_val_adj < .05)
AGPP_KEGG_Markers
AGPP_KEGG_Markers$gene <- rownames(AGPP_KEGG_Markers)
AGPP_KEGG_DE_genes <- AGPP_KEGG_Markers$gene
AGPP_KEGG_DE_genes <- sort(AGPP_KEGG_DE_genes)
AGPP_KEGG_DE_genes
AGPP_KEGG_DE_gene_index <- c(1,2,1,1,1,2,2,2,2,2,2,2,2,2,2,1,1,3,3,3,3)
AGPP_KEGG_DE_genes <- AGPP_KEGG_DE_genes[order(AGPP_KEGG_DE_gene_index)]

DefaultAssay(CD_EGCs.integrated) <- "RNA"
CD_EGCs.genes <- rownames(CD_EGCs.integrated)
CD_EGCs.integrated <- ScaleData(CD_EGCs.integrated, features = CD_EGCs.genes)

CD_EGCs.integrated$Condition <- factor(
CD_EGCs.integrated$Condition,
levels = c("Healthy", "Crohn's"))

CD_EGC_AGPP_Heatmap <- DoHeatmap(CD_EGCs.integrated, features = AGPP_KEGG_DE_genes, group.by = "Condition", label = F, group.colors = c("#232D4B", "#E57200")) + theme(axis.text.y = element_text(face = "bold.italic", size = 20), legend.text = element_text(size = 20), legend.title = element_text(size = 20)) + guides(color = "none")
CD_EGC_AGPP_Heatmap
```